### PR TITLE
Add support for BCEWithLogitsLoss

### PR DESCRIPTION
**Pull request by baijumeswani**
This pull request adds support for composing an onnx model with the `BCEWithLogitsLoss` loss. Example usage of this loss is as:

```py
class MyModel(onnxblock.TrainingModel):
    def __init__(self):
        super(MyModel, self).__init__()
        self.loss = onnxblock.loss.BCEWithLogitsLoss()

    def build(self, output_name):
        return self.loss(output_name)

model = MyModel()
with onnxblock.onnx_model(base_onnx_model) as accessor:
    loss_output_names = model()
    eval_model = accessor.eval_model
```

In addition to the support for `BCEWithLogitsLoss` above, this pull request also introduces a concept of a `Block`. A `Block` is the base class for all other onnx blocks (such as `Sub`, `Add`, `Sigmoid`.. as well as the `onnxblock.Model` and `onnxblock.TrainingModel`) that can be used to stack layers of the onnx model.

The distinction between a `Block` and a `Model` is that the `Block` does not bother itself with graph output generation. It only connects components together. The `Model` on the other hand is the user facing class that is responsible for shape inferencing and graph output generation.

### CODE REVIEW

**Review by pengwa — APPROVED** [review]
LGTM

File: orttraining/orttraining/python/training/onnxblock/building_blocks.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            node_input_names,
+            node_output_names,
+            name=graph_utils.generate_random_graph_name(op_name),
+        )
+        onnx_model.graph.node.append(node)
+
+        return node_output_name
+
+
+class Add(_BinaryOp):
```
Comment (askhade) [inline orttraining/orttraining/python/training/onnxblock/building_blocks.py:63]: Do we need a separate class for each binary op? Can we simply call the build mtd of BinaryOp with the op_name?

File: orttraining/orttraining/python/training/onnxblock/_graph_utils.py [hunk 2]
```diff
@@ -203,3 +201,56 @@ def get_model_parameters(model, args_not_requiring_gradient):
...
+    existing_graph_outputs = {output.name for output in model.graph.output}
+
+    if isinstance(output_names, str):
+        output_names = [output_names]
+
+    # collect all new graph outputs (i.e. graph outputs that are not
+    # already graph outputs)
+    graph_output_names_to_build = []
+    for output_name in output_names:
+        if output_name not in existing_graph_outputs:
```
Comment (askhade) [inline orttraining/orttraining/python/training/onnxblock/_graph_utils.py:246]: if override_outputs is set to True then according to current code: we will not include the output names present in output_names if they are present in existing_graph_outputs... Not sure if this is intended

File: orttraining/orttraining/python/training/onnxblock/building_blocks.py (hunk 1 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/building_blocks.py:63]: I can potentially make this simpler, but I don't want users to instantiate an instance of `BinaryOp` using `add = BinadryOp("Add")`. This kind of instantiation in my opinion is not very programmatic and can lead to more string mistyping errors.
I think it is better to have a class for each class.

File: orttraining/orttraining/python/training/onnxblock/_graph_utils.py (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/_graph_utils.py:246]: Yeah, the intention of `override_outputs` was to clear the existing graph outputs and make room for new ones. And I later added the case where users might want to append to existing outputs.
I think this should be made more generic to mean that if override_outputs is True, then the outputs should be all the outputs in `output_names`. Else, it should be existing graph outputs+output_names. I'll make that correction.

File: orttraining/orttraining/python/training/onnxblock/_graph_utils.py [hunk 3]
```diff
@@ -203,3 +201,40 @@ def get_model_parameters(model, args_not_requiring_gradient):
                 trainable_params.append(initializer)
 
     return trainable_params, non_trainable_params
+
+
+def build_graph_outputs(model, output_names):
+    """Append graph outputs for the model for the given output names.
+
+    The graph outputs are extracted from the graph value_infos and
```
Comment (askhade) [inline orttraining/orttraining/python/training/onnxblock/_graph_utils.py:209]: graph.value_info contains value info for intermediate outputs ... basically after shape inference value info for all the intermediate outputs  is placed in value_info and all the graph outputs is placed in output field.

**Review by askhade — APPROVED** [review]
LGTM

**Comment by baijumeswani** [comment]
Thanks for the reviews. :)
